lib/runner.js

```diff
@@ -0,0 +1,52 @@
+"use strict";
+
+module.exports = class SerializedRunner {
+	constructor(asyncOp) {
+		this.asyncOp = asyncOp;
+	}
+
+	run() {
+		if(!this._pending) { // prevent concurrent execution
+			this._pending = augment(this.asyncOp()).
+				finally(_ => {
+					this._pending = null;
+				});
+		}
+		return this._pending;
+	}
+
+	// repeats execution exactly once after waiting for any pending execution to conclude
+	rerun() {
+		if(this._queued) { // limit queue to a single repeat execution
+			return this._queued;
+		}
+
+		let enqueue = this._pending;
+		let res = this.run();
+		if(enqueue) {
+			this._queued = res = augment(res).
+				finally(_ => {
+					this._queued = null;
+				}).
+				then(_ => this.run());
+		}
+		return res;
+	}
+};
+
+function augment(promise) {
+	promise.finally = always;
+	return promise;
+}
+
+// poor man's `Promise#finally` polyfill
```

Comment:
That's a ponyfill, young man.

Reply:
Nope; the ponyfill is `augment` above... (I expended an inordinate amount of effort pondering that.)